Add product review title and rating blocks

louwie17 opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR builds on top of the initial product reviews PR ( #56276 ) and adds two new blocks:
- **Product Review Rating** - A block to display the **star** rating of an individual review.
- **Product Reviews Title** - A block to replace the **Post Comments Title**, which updates the wording to use reviews, but also makes use of the WC Reviews REST endpoint within context aware blocks.

Part of https://github.com/woocommerce/woocommerce/issues/55729

Closes WOOPLUG-3849

### Screenshots or screen recordings:

<img width="1512" alt="Screenshot 2025-03-13 at 1 44 09 PM" src="https://github.com/user-attachments/assets/c16d54fd-6056-4640-8a07-1a6b19eab9c4" />

https://github.com/user-attachments/assets/54af7715-51aa-4262-aaeb-fc781981b59a

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

**Note:** These tests will be focused towards the two new blocks, the reviews title and the review rating. Meaning that the review block still contains other blocks that pertain to comments specifically. These will be addressed in follow up PRs.

1. Load this branch and have the **WooCommerce Beta Tester** plugin activated
2. Go to **Tools > WCA Test Helper > Features** and enable the **experimental blocks** feature flag
3. Go to **Appearance > Editor > Templates** and edit the **Single Product** template
4. Add the **Blockified Product Details** block ( an accordion )
5. The third accordion ( Reviews ) should contain the **Blockified Product Reviews** block as its child
6. It should include the **Product Reviews Title** as the first nested block that uses the "product reviews" copy and allows you toggle the review count and product title ( this should also be reflected correctly on the front-end ).
7. The comment template will have a third column that contains the product rating ( the 5 stars )
8. Save the template
9. Visit a product with and without reviews, if reviews exist, it should correctly render the title with the correct number and show the correct star rating for each review, if no reviews exist, it shouldn't render any review and render "0 reviews for ..." as the title. ( We may want to create a follow up to have an explicit "No review" title block, as in the current block it displays "There are no reviews yet." and "Be the first to review “Black and White”" )

**Test areas with context**

Follow the tests below either within a page where you use either the **Single product** or **product collection** block.

1. Add the **Blockified Product Reviews** block as a child of one of the above blocks ( make sure you have a product with and without reviews displayed )
2. If the parent block has reviews it should show the correct number in the **Product Reviews Title** block. Ratings will unfortunately not be shown yet, because we still make use of the **Comments Template** which uses the comments endpoint.
3. If the parent block does not have reviews, it should show the correct number of zero reviews in the title.
4. Save the page and preview it, it should render the title correctly and also the reviews that include the star rating.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

louwie17 committed:
Add reviews block

louwie17 committed:
Add Product Review Title block

louwie17 committed:
Fix styling

louwie17 committed:
Add changelog

louwie17 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/block.json:
```diff
@@ -0,0 +1,24 @@
+	"category": "woocommerce",
+	"description": "Displays the rating of a product review.",
+	"textdomain": "woocommerce",
+	"usesContext": [ "commentId" ],
```
Sticking with `commentId` given we will keep the comments template component that exposes the comment id.

louwie17 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-title/edit.tsx:
```diff
@@ -0,0 +1,170 @@
+			return;
+		}
+		resolveSelect( reviewsStore )
+			.getReviewsTotalCount( {
```
Had to use the `reviewsStore` for the data as the WP comments API doesn't work for this.
This is the main reason I created a new block for this.

louwie17 commented on plugins/woocommerce/src/Blocks/BlockTypes/Reviews/ProductReviewsTitle.php:
```diff
@@ -0,0 +1,98 @@
+	 * @param array $attributes Block attributes.
+	 * @return string Rendered block content.
+	 */
+	protected function render( $attributes, $content, $block ) {
```
This is based off of the comments-title render function, just making use of the actual `$product` entity.

louwie17 committed:
Fix lint error

louwie17 committed:
Fix lint errors

louwie17 committed:
Fix lint errors

imanish003 reviewed (commented):
Hey @louwie17, I've completed functional testing of the PR and identified a few minor issues that I'll detail below. Thanks for your work on this!

- Verified that third accordion contains the Blockified Product Reviews block as its child
- Verified that the Product Reviews Title block is present and displays text `3 reviews for "Product Title"`
- Toggles "Show Reviews Count" and "Show Product Title" correctly update the title in Editor side as well as on the front-end

> We may want to create a follow up to have an explicit "No review" title block

Makes sense. In my testing, when there are no reviews:
- And if Product Title is hidden, it shows "0 reviews" on Frontend
- When Reviews Count is hidden, it shows `Reviews for “Polo”` on Frontend

Which I agree, isn't best UX. 

> as in the current block it displays "There are no reviews yet." and "Be the first to review “Black and White”"

This doesn't seem to be the case for me, Am I missing something? Here is a screenshot of what I am seeing on Frontend:

![image](https://github.com/user-attachments/assets/5d9120fc-62c1-497b-93c4-73f75b2b555f)

#### Minor: Star rating is cut off

As you can see in the screenshot below, the stars are cut off or truncated. Are you also seeing this?
![image](https://github.com/user-attachments/assets/55b9d56f-ebe5-407c-8107-6e700add5d26)

#### Minor: Accordions are appearing and disappearing on Frontend

I don't think it is directly related to this PR but I noticed on front end that at first all the accordions are appearing and after some time only one accordion remains which is description accordion. I was able to reproduce it for an album product but when I tried to reproduce it for a cap product, I was not able to reproduce it. So I am not sure if there is something wrong with my local setup, but in any case, I thought it's worth mentioning.

https://github.com/user-attachments/assets/ca8a109e-ba01-47de-8517-7574d3942484

### Unexpected behavior when adding this block to a product collection block

The video below demonstrates an unexpected behavior when adding this block to a product collection block. In the editor, when I add the block to a product collection block, only one product displays on the frontend if the block is part of the product template. Maybe there is a bug in Product Collection block? Are you able to reproduce this?

Sidenote: This seems to be working fine inside Single Product block.

https://github.com/user-attachments/assets/37ce9bb2-b551-43f0-a147-b7a3c2e14f43

louwie17 committed:
Update template structure to fix review rating overlap

louwie17 committed:
Disable front end script for review blocks

louwie17 commented:
Thanks for the review @imanish003 :) 

> This doesn't seem to be the case for me, Am I missing something? Here is a screenshot of what I am seeing on Frontend:

From the image it looks like your looking at the new block, when mentioning the `current block` here "current block it displays "There are no reviews yet." and "Be the first to review “Black and White”"" I was referring to the current non-blockified reviews version:
<img width="634" alt="Screenshot 2025-03-19 at 09 20 57" src="https://github.com/user-attachments/assets/29a89c9d-e867-4df6-bbc8-7e23a44e10bb" />

> #### Minor: Star rating is cut off
> As you can see in the screenshot below, the stars are cut off or truncated. Are you also seeing this?

I fixed this as part of this commit: https://github.com/woocommerce/woocommerce/pull/56332/commits/7164e9988063801404c9bb58ae0c4093dc341800, as it was because of the fixed width I set the third column to, which wasn't big enough for the rating.
In hindsight, turning it into three columns is not the right approach, what I did instead is I moved the rating and the comment author block into a `row` block with justification set to "Space inbetween items". This way the star review has plenty of room and it should also scale better. 

> #### Minor: Accordions are appearing and disappearing on Frontend
> I don't think it is directly related to this PR but I noticed on front end that at first all the accordions are appearing and after some time only one accordion remains which is description accordion. I was able to reproduce it for an album product but when I tried to reproduce it for a cap product, I was not able to reproduce it. So I am not sure if there is something wrong with my local setup, but in any case, I thought it's worth mentioning.

I cannot reproduce this! What theme are you using? This kinda smells like a styling issue with the theme?

> ### Unexpected behavior when adding this block to a product collection block
> The video below demonstrates an unexpected behavior when adding this block to a product collection block. In the editor, when I add the block to a product collection block, only one product displays on the frontend if the block is part of the product template. Maybe there is a bug in Product Collection block? Are you able to reproduce this?

Interesting!! So I did some further digging and this turns out to be an issue with the product collection block I believe in particular when using a comment form.
If I create the same thing but in a query loop, it renders fine!
It also is not strictly related to the **Blockified product reviews** block, and just adding the **Comments form** block only within the product collection also triggers the issue.

I don't know if the product collection block does some parsing of the HTML? but I have a suspicion that somewhere some HTML is being modified and a closing tag is being removed. Which in turn causes some of the nested HTML to disappear and the interactivity API not being able to find some elements.

louwie17 commented:
@imanish003 this should be ready for a re-review given the product collection issue is not related to this PR, and I addressed the rating overlap issue.

imanish003 commented:
> From the image it looks like your looking at the new block, when mentioning the current block here "current block it displays "There are no reviews yet." and "Be the first to review “Black and White”"" I was referring to the current non-blockified reviews version:

Ah! My bad 🤦‍♂️

> I fixed this as part of this commit: https://github.com/woocommerce/woocommerce/commit/7164e9988063801404c9bb58ae0c4093dc341800, as it was because of the fixed width I set the third column to, which wasn't big enough for the rating.
> In hindsight, turning it into three columns is not the right approach, what I did instead is I moved the rating and the comment author block into a row block with justification set to "Space inbetween items". This way the star review has plenty of room and it should also scale better.

Great! It's working as expected now.

> I cannot reproduce this! What theme are you using? This kinda smells like a styling issue with the theme?

This is occurring for certain products such as the album and Hoodie with Logo, as shown in the demo video below. It is not reproducible across all products. Here are some environmental details in case they are helpful.

WC Version: 9.9.0-dev
WP Version: 6.7.2
Twenty Twenty-Five


https://github.com/user-attachments/assets/c21f24db-9c38-4d42-b55e-ae9a0938879d

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/edit.tsx:
```diff
@@ -0,0 +1,94 @@
+import { useStyleProps } from '@woocommerce/base-hooks';
+import type { BlockEditProps } from '@wordpress/blocks';
+
+/**
+ * Renders the `woocommerce/product-review-rating` block on the editor.
+ *
+ * @param   {Object}  props                    React props.
+ * @param   {Object}  props.attributes         Block attributes.
+ * @param   {Object}  props.setAttributes      Function to set block attributes.
+ * @param   {Object}  props.context           Inherited context.
+ * @param   {string}  props.context.commentId The comment ID.
+ *
+ * @return {JSX.Element} React element.
+ */
+export default function Edit( {
```
I see an ESLint warning. Do you see it too?

![image](https://github.com/user-attachments/assets/705bf045-3abd-482a-8c4c-cbfb8b05e434)

louwie17 replied:
Yeah I got the same thing, I will remove it actually, as it is not common for us to add these JS docs to begin with and the component is also typed already.
I had initially mimicked this from the GB approach I guess.

louwie17 replied:
Removed the comment block as part of this commit: https://github.com/woocommerce/woocommerce/pull/56332/commits/ab16e279c648dd788b7cafea8172c02e9538e9fa

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/edit.tsx:
```diff
@@ -0,0 +1,94 @@
+	context: { commentId: string };
+} ) {
+	const { textAlign } = attributes;
+	const styleProps = useStyleProps( attributes );
```
Out of curiosity, do we really need to use `useStyleProps` here? Isn't there any alternative present in Gutenberg APIs? For example - [__experimentalGetColorClassesAndStyles](https://github.com/WordPress/gutenberg/blob/1c698f8ccdcc8a86b91f9e7efef608ad66f3a46b/packages/block-editor/src/index.js#L8). TBH I have not looked into Gutenberg APIs for this but I am wondering if there is an API in Gutenberg which we can use for this purpose, instead of custom API.

louwie17 replied:
Removed it: https://github.com/woocommerce/woocommerce/pull/56332/commits/ab16e279c648dd788b7cafea8172c02e9538e9fa :)

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/edit.tsx:
```diff
@@ -0,0 +1,94 @@
+	const styleProps = useStyleProps( attributes );
+	const className = clsx(
+		styleProps.className,
+		'wc-block-components-product-review-rating',
```
Nit: Is there any reason to have components in this className? 

```suggestion
		'wc-block-product-review-rating',
```

louwie17 replied:
Nope, removed it: https://github.com/woocommerce/woocommerce/pull/56332/commits/ab16e279c648dd788b7cafea8172c02e9538e9fa

louwie17 commented:
> This is occurring for certain products such as the album and Hoodie with Logo, as shown in the demo video below. It is not reproducible across all products. Here are some environmental details in case they are helpful.
> 
> WC Version: 9.9.0-dev WP Version: 6.7.2 Twenty Twenty-Five

@imanish003 unfortunately this is not super helpful! Are you able to dig in a bit deeper on your end to see what might be going on? For example are the actual accordions disappearing? or is there a style that is changing that is overlapping the two accordions? Are you able to reproduce this on a fresh site? Is this only happening for you on this branch or also on `trunk`?
What happens if you move the **blockified product details** block down below the **columns** or **Product Details** block, is the same thing happening?

We could also jump on a call.

Btw I am also using the twenty twenty-five theme in the video below on a fresh site and it looks a lot different than yours, especially the classic product reviews block. Are you sure your not importing any other styles or have other plugins enabled? Notice how yours has the borders for the classic reviews block and mine doesn't:

https://github.com/user-attachments/assets/31877d4b-1e32-4e58-bee4-cc97ccdccb4e

> Is this only happening for you on this branch or also on `trunk`?

If this is also happening on `trunk` and every thing else tests well, could we approve this PR? and continue debugging the above issue and fix it in a separate PR?

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/edit.tsx:
```diff
@@ -0,0 +1,94 @@
+			<div { ...blockProps }>
+				<div
+					className={
+						'wc-block-components-product-review-rating__stars'
```
```suggestion
						'wc-block-product-review-rating__stars'
```

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/edit.tsx:
```diff
@@ -0,0 +1,94 @@
+	const blockProps = useBlockProps( {
+		className,
+	} );
+	let [ rating ] = useEntityProp( 'root', 'comment', 'rating', commentId );
+	rating = rating ?? 4;
```
qq: Will we refactor this code to be review-specific or will this remain same?

louwie17 replied:
Which code are you referring to specifically?
We will continue to use the `useEntityProp` with the `commentId` as context, given we use the same context as part of the variations from the comment blocks.
Currently the above works fine if we provide a review id, and its not supported to use `useEntityProp( 'root', 'product-review', 'rating', commentId );` for example.

I may explore adding the support in the review data store for retrieving a single review https://woocommerce.github.io/woocommerce-rest-api-docs/#retrieve-a-product-review and use that instead.

louwie17 replied:
I end up adding the `getReview` selector to the reviews data store and used that here instead of `useEntityProp`: https://github.com/woocommerce/woocommerce/pull/56332/commits/540e2ff7da385331ebedd2297b4b704daf518d41

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/index.tsx:
```diff
@@ -0,0 +1,27 @@
+
+if ( isExperimentalBlocksEnabled() ) {
+	// @ts-expect-error metadata is not typed.
+	registerBlockType( metadata, {
```
I think we can use `registerProductBlockType` here as well. Alternatively, we can set `"ancestor": [ "woocommerce/blockified-product-reviews" ],` property.

louwie17 replied:
I made use of `ancestor` -> https://github.com/woocommerce/woocommerce/pull/56332/commits/68afb7b637ed741782dd89e04ef58e08f4c674ea

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/review-rating/index.tsx:
```diff
@@ -0,0 +1,27 @@
+	// @ts-expect-error metadata is not typed.
+	registerBlockType( metadata, {
+		edit,
+		icon: {
+			src: (
+				<Icon
+					icon={ starHalf }
+					className="wc-block-editor-components-block-icon"
+				/>
+			),
+		},
```
FWIW, out of curiosity, I checked whether directly passing the icon works, and it seems that it does. I was just curious and looked at the [code in Gutenberg](https://github.com/WordPress/gutenberg/blob/1c698f8ccdcc8a86b91f9e7efef608ad66f3a46b/packages/blocks/src/api/utils.js#L94-L130).

```suggestion
		icon: (
			<Icon
				icon={ starHalf }
				className="wc-block-editor-components-block-icon"
			/>
		),
```

louwie17 replied:
Yeah even the use of `Icon` is not necessary, and we can just use `icon: starHalf`.
I don't quite remember doing this explicitly myself, so I may of copied this from the average rating block: https://github.com/woocommerce/woocommerce/blob/99f04562c084808e02515a526664fdc0c77fa688/plugins/woocommerce/client/blocks/assets/js/atomic/blocks/product-elements/average-rating/index.tsx#L16

louwie17 replied:
Just passed icon directly: https://github.com/woocommerce/woocommerce/pull/56332/commits/1b60ec1bb62284ce683c744014552a7006ef5a77

imanish003 replied:
Nice 🚀

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-title/edit.tsx:
```diff
@@ -0,0 +1,170 @@
+	);
+
+	const postTitle = isSiteEditor
+		? __( '“Product Title”', 'woocommerce' )
```
```suggestion
		? __( '"Product Title"', 'woocommerce' )
```

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-title/edit.tsx:
```diff
@@ -0,0 +1,170 @@
+		</InspectorControls>
+	);
+
+	const postTitle = isSiteEditor
```
```suggestion
	const productTitle = isSiteEditor
```

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-title/edit.tsx:
```diff
@@ -0,0 +1,170 @@
+				placeholder = sprintf(
+					/* translators: 1: Number of comments, 2: Post title. */
+					_n(
+						'%1$s review for %2$s',
+						'%1$s reviews for %2$s',
+						reviewsCount,
+						'woocommerce'
+					),
+					reviewsCount,
+					postTitle
+				);
+			}
+		} else if ( reviewsCount === 1 ) {
+			placeholder = __( 'One review', 'woocommerce' );
+		} else {
+			placeholder = sprintf(
+				/* translators: %s: Number of reviews. */
+				_n( '%s review', '%s reviews', reviewsCount, 'woocommerce' ),
+				reviewsCount
+			);
+		}
+	} else if ( showProductTitle ) {
+		if ( reviewsCount === 1 ) {
+			placeholder = sprintf(
+				/* translators: %s: Post title. */
+				__( 'Review for %s', 'woocommerce' ),
+				postTitle
+			);
+		} else {
+			placeholder = sprintf(
+				/* translators: %s: Post title. */
+				__( 'Reviews for %s', 'woocommerce' ),
+				postTitle
+			);
+		}
+	} else if ( reviewsCount === 1 ) {
+		placeholder = __( 'Review', 'woocommerce' );
+	} else {
+		placeholder = __( 'Reviews', 'woocommerce' );
+	}
```
qq: Is it possible to avoid nested if-else structures and instead use a flatter if-else structure? I find this code a bit harder to understand. Maybe it's just me who feels this way 😅

louwie17 replied:
Hmm while I see your point, I am afraid it may not help, and if we do flatten the structure, we end up with a lot of large if statements. I also debated on doing some concatenation fun, but that would mess up the translation strings.

Btw I didn't write these if statements myself: https://github.com/WordPress/gutenberg/blob/317c348874fb5d4cf47297351aee4c9098cf15c6/packages/block-library/src/comments-title/edit.js#L138

louwie17 replied:
Let me move the logic to its own function actually and make use of early return statements, those may help

louwie17 replied:
I moved it to a function and made use of the early return statements to flatten the conditions, I did it both for the JS version and the PHP version: https://github.com/woocommerce/woocommerce/pull/56332/commits/346c7250db33f3a7958ed644dfa206f18c99f402
I hope this helps :)

imanish003 replied:
That looks much better to me. Thanks a lot for making this change 🙇

imanish003 commented on plugins/woocommerce/src/Blocks/BlockTypes/Reviews/ProductReviewRating.php:
```diff
@@ -0,0 +1,85 @@
+
+		$rating = intval( get_comment_meta( $block->context['commentId'], 'rating', true ) );
+
+		$classes_and_styles = StyleAttributesUtils::get_classes_and_styles_by_attributes( $attributes, array(), array( 'extra_classes' ) );
```
Is it necessary to use `StyleAttributesUtils`? Is it possible to use [get_block_wrapper_attributes](https://github.com/woocommerce/woocommerce/issues/51648) here? TBH I haven't looked deeply into it but I remembered that we were discussing about moving away from this.

louwie17 replied:
Nice catch! This was also copy/paste from the `ProductAverageRating` class 🤦‍♂️

louwie17 replied:
Removed: https://github.com/woocommerce/woocommerce/pull/56332/commits/e1ea5daa95f09cb62bbc5593cd86dd295d691a60

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-title/style.scss:
```diff
@@ -0,0 +1,4 @@
+
+.wp-block-woocommerce-product-reviews-title.has-background {
```
Should we consider using a class with a WC prefix instead of a WP prefix here? This way, when third-party developers add some styles, they can rely on this specific class for this block?

```suggestion
.wc-block-product-reviews-title.has-background {
```

louwie17 replied:
Hmm is this something we do in other places, or do we have some logic that auto adds a `wc-block` generated class?
To be honest, on order for me to do the above I have to add the `wc-block-*` class specifically in multiple places, I rather rely on the auto generated `wp-block` class that is based off of the block name.

imanish003 replied:
I see and I somewhat agree with you that it might be better to rely on the auto-generated class instead of specifying the class with the WC prefix in multiple places, as that could lead to more errors. I also asked about this on Slack (p1742547412664939-slack-C02FL3X7KR6).

imanish003 replied:
It seems that we have reached a consensus on Slack to use the auto-generated class. Therefore, I am resolving this comment.

imanish003 commented on plugins/woocommerce/client/blocks/assets/css/style.scss:
```diff
@@ -99,10 +99,12 @@ body.wc-block-product-gallery-modal-open {
 
-.wc-block-grid__product-rating {
+.wc-block-grid__product-rating,
+.wp-block-woocommerce-product-review-rating {
```
Similar to the other comment, Should we consider using a class with a WC prefix instead of a WP prefix here?

louwie17 replied:
I replied above, but in my opinion "no". Unless we have a specific guideline that we follow to do this for all our WooCommerce blocks.
If not, I think relying on the block generated name is just as fine.

imanish003 reviewed (changes requested):
@louwie17, you’ve done an amazing job on this PR! Everything looks really good overall. I’ve shared a few minor comments and suggestions for you to consider. I’d love to hear your thoughts on them, and once we go over those little details, I’ll be happy to approve the PR! 🚀

louwie17 committed:
Updated class name and removed unneeded style props

louwie17 committed:
Add getReview selector and make use of this

louwie17 committed:
Add ancestor to product review rating block

louwie17 committed:
Update icon usage of review rating block

louwie17 committed:
Simplify the reviews title logic

louwie17 committed:
Remove unneeded use of StyleAttributesUtils

louwie17 commented:
@imanish003 thanks for the detailed review, I addressed your suggestions and left a couple replies to some of your other questions, in particular around the `wc-block-*` class.

louwie17 committed:
Fix css lint errors

louwie17 committed:
Add data package changelog

imanish003 commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-reviews/inner-blocks/reviews-title/edit.tsx:
```diff
@@ -0,0 +1,187 @@
+	if ( showReviewsCount && showProductTitle ) {
+		return reviewsCount === 1
+			? sprintf(
+					/* translators: %s: Post title. */
```
Nit: Can we update this in all places? 

```suggestion
					/* translators: %s: Product title. */
```

louwie17 replied:
Done, thanks: https://github.com/woocommerce/woocommerce/pull/56332/commits/f643843317f856cfad884a280fbb0fb194d3d1c7

imanish003 commented on packages/js/data/src/reviews/resolvers.ts:
```diff
@@ -36,6 +36,23 @@ export function* getReviews( query: ReviewsQueryParams ) {
 
+export function* getReview( id: number ) {
+	try {
+		const url = addQueryArgs( `${ NAMESPACE }/products/reviews/${ id }` );
```
I checked, and here `NAMESPACE` points to `/wc-analytics`. Is there a reason we rely on this endpoint? As far as I understand, this endpoint is specifically for views related to analytics. Should we consider using `wc/v3` instead? I’m new to this part of the codebase, so I may have misunderstood something.

louwie17 replied:
Yeah I did notice the same thing, but turns out that the `wc-analytics` makes use of the REST Api v3 class in behind the scenes: https://github.com/woocommerce/woocommerce/blob/3dd60322d5490d7d86ad33e7d732f276df73a12d/plugins/woocommerce/src/Admin/API/ProductReviews.php#L18

Having said that it is probably worth refactoring this, so the review class uses `wc/v3` always and we make an exception for the analytics endpoint.
I would suggest I do this as a follow up, to unblock this work.

I did remove the use of namespace for now: https://github.com/woocommerce/woocommerce/pull/56332/commits/ea5ea6d0292585e884a044844cbca4be966ad224 and will update the `getReviews` function in a follow up.

imanish003 replied:
I see! FWIW there is also a feature flag, and based on that flag's value, [some of the analytics API may not be loaded](https://github.com/woocommerce/woocommerce/blob/ea5ea6d0292585e884a044844cbca4be966ad224/plugins/woocommerce/src/Admin/API/Init.php#L124). However, it looks like the product reviews analytics API will always be loaded. So yeah, I think we can merge this as it is, and as you mentioned, you could maybe follow up in a new PR.

> I did remove the use of namespace for now: https://github.com/woocommerce/woocommerce/commit/ea5ea6d0292585e884a044844cbca4be966ad224 and will update the getReviews function in a follow up.

Nice! 🚀

imanish003 reviewed (changes requested):
Thanks for making all the changes. Everything looks good to me now, except for a few more minor things—especially my comment about `NAMESPACE` constant.

louwie17 committed:
Update translator comments

louwie17 committed:
Use wc/v3 endpoint instead

louwie17 commented:
@imanish003 should be good for a re-review, I also replied to your question around the analytics endpoint usage. I think we should update this as a follow up. The endpoint uses the `wc/v3` class in behind the scenes and only updates the prepare links as the difference: https://github.com/woocommerce/woocommerce/blob/3dd60322d5490d7d86ad33e7d732f276df73a12d/plugins/woocommerce/src/Admin/API/ProductReviews.php#L32

I initially debated on creating a new data store that uses the `wc/v3` endpoint only, but that would add additional confusion having two review data stores. I think this store should be updated to use `wc/v3` by default and use the analytics endpoint if needed.
But as suggested in the comment I think this should be a follow up.

imanish003 commented:
> @imanish003 should be good for a re-review, I also replied to your question around the analytics endpoint usage. I think we should update this as a follow up. The endpoint uses the `wc/v3` class in behind the scenes and only updates the prepare links as the difference:
> 
> https://github.com/woocommerce/woocommerce/blob/3dd60322d5490d7d86ad33e7d732f276df73a12d/plugins/woocommerce/src/Admin/API/ProductReviews.php#L32
> 
> I initially debated on creating a new data store that uses the `wc/v3` endpoint only, but that would add additional confusion having two review data stores. I think this store should be updated to use `wc/v3` by default and use the analytics endpoint if needed. But as suggested in the comment I think this should be a follow up.

Yeah, this makes sense to me. I am going to do final testing before approving the PR.

imanish003 commented on plugins/woocommerce/src/Blocks/BlockTypes/Reviews/ProductReviewsTitle.php:
```diff
@@ -0,0 +1,128 @@
+			$tag_name = 'h' . $attributes['level'];
+		}
+
+		if ( 0 === $reviews_count ) {
+			return;
+		}
```
I just wanted to confirm whether this is intentional because, according to the PR description, it should display text that says "0 reviews for...". However, based on this condition, it won't show anything at all on the front end. So, I wanted to clarify if this is intentional or not.

imanish003 replied:
If this is intentional, I believe we (should?) also update the editor and shouldn't display anything on the editor side. Here is a screenshot comparing the editor and the front endview.
![image](https://github.com/user-attachments/assets/be015844-d8a3-4024-a405-20b284bf5594)

louwie17 replied:
Good call out, I decided to still show the title on the front end as well https://github.com/woocommerce/woocommerce/pull/56332/commits/a46845693705fde017a6d9b88117a4de23b7fd92, I had mimicked this logic from the core comments title block as well.
The problem is that you can't really hide the title in the block editor, or I am not aware of a good way, that would still allow you to edit it.
If we do want to have the option to hide it, we could add a setting to "hide title if no reviews", we can keep that conversation going as a follow up. In either case things should be identical now.

@j111q & @ellenbauer do you think we need a setting for the reviews title to give users the option to hide it if there are 0 reviews? Or should we always show a title?

imanish003 replied:
Sure, it makes sense to continue the discussion and, if needed, follow up with the PR. FWIW I wanted to point out that in the screenshot above, the ratings block is showing "no reviews" on the editor side, and on the front end, it is not rendering anything at all similar to this block.

j111q replied:
I actually wonder if we should NEVER show a title, for any of the accordion sections -- meaning we don't show the "Description" title or "Additional info" or "Reviews" title, ever. Because the Accordion title itself already serves as a title, if you get what I mean. 

Here's some nice Woo stores as an example:
- https://bario-neal.com/product/rittenhouse-chain/
- https://shop.thecoolhunter.net/product/silver-cocktail-coasters/

imanish003 reviewed (approved):
Hey @louwie17, thank you so much for your patience and amazing work on this PR. I know there has been a lot of back and forth, and I truly appreciate your patience. Everything looks good to me now. I have tested the PR again, and everything is working as expected. I left one minor comment, but I am going to pre-approve the PR. I think once we address that comment, we can merge this PR. 🚀

louwie17 committed:
Still show title if no reviews
